File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bruteforce/Problem15652.java
package com.keunwon.algorithm.bruteforce;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.util.Scanner;

public class Problem15652 {
    public static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static int[] arr;

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);

        final int N = sc.nextInt();
        final int M = sc.nextInt();
        arr = new int[M];

        dfs(N, M, 0, 0);

        bw.flush();
        bw.close();
    }

    public static void dfs(int n, int m, int start, int depth) throws IOException {
        if (m == depth) {
            for (int i : arr) {
                bw.write(i + " ");
            }
            bw.newLine();
            return;
        }

        for (int i = start; i < n; i++) {
            arr[depth] = i + 1;
            dfs(n, m, i, depth + 1);
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem15652.kt
package com.keunwon.algorithm.baekjoon

class Problem15652 {
    private val answer = mutableListOf<String>()

    fun solution(n: Int, m: Int): List<String> {
        dfs(0, 1, IntArray(m), n)
        return answer
    }

    private fun dfs(depth: Int, startIndex: Int, picks: IntArray, n: Int) {
        if (depth == picks.size) {
            answer.add(picks.joinToString(" "))
            return
        }

        for (i in startIndex..n) {
            picks[depth] = i
            dfs(depth + 1, i, picks, n)
        }
    }
}

fun main() = with(System.`in`.bufferedReader()) {
    val bw = System.out.bufferedWriter()
    val (n, m) = readLine().split(" ").map { it.toInt() }

    Problem15652().solution(n, m).forEach {
        bw.write(it)
        bw.newLine()
    }

    bw.flush()
    bw.close()
    close()
}
